Proposal: Asynchronous Listener Registration (#989)
hober/web-extensions: proposals/async-listener-registration.md

# Proposal: Asynchronous Listener Registration

**Summary**

> An API proposal to allow extensions to explicitly manage their service worker initialization phase via a manifest opt-in field and a `markInitializationComplete()` API, enabling the asynchronous registration of event listeners. 

Reviewer: "manage their service worker initialization" sounds broad (potentially raising concerns about impact on service workers), but my impression after reading the whole document is that this is really about delaying all extension API event dispatch at the start of the extension service worker wakeup, until the extension signals that it ready to receive (or drop) them.

Notably out of scope are: standard SW lifecycle events, web platform APIs events, non-event.

Is my interpretation accurate? If so, can we make this more explicit in the description and choice of the API names (I'm not suggesting names yet, bikeshedding can be done once we have a common understanding of its semantics).

E.g.

```suggestion
An API proposal to allow extensions to explicitly delay the dispatch of extension API events to their service worker during its initialization phase via a manifest opt-in field and a `markInitializationComplete()` API, enabling the asynchronous registration of event listeners. 
```

P.S. "extension service worker" is not supported by Firefox yet, the predecessor (event pages) are still supported by Firefox and Safari ( https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support ).
Where I read "extension service worker" across the proposal, I have the understanding that it refers to this broader concept (perhaps it makes sense to write that in a small note somewhere, for those not in the loop).

Author: Yes, your interpretation is accurate. I'll reword it to clarify that this only delays dispatch of extension API events to the background context during startup.

I’ll use "background context" to cover extension service workers and event pages.

If an extension registers a listener asynchronously (e.g., inside a `setTimeout`, after a `fetch()`, or after reading from `browser.storage`), the initial script evaluation finishes before the listener is attached. Consequently, when the relevant event fires later, the service worker wakes up, executes its top-level script synchronously, and the browser dispatches the event to all registered listeners (of which there may be none). Thus, the service worker wakes up but receives no event.

This proposal introduces a mechanism to explicitly delay the finalization of the "initialization phase" of the worker and formalizes how listener state is managed and persisted across service worker restarts.

#### Use Cases

* **Conditional event registration:** An extension may need to read user preferences from local storage to determine whether it should register listeners for high-traffic events (e.g., `tabs.onUpdated` or `webRequest`).
* **Dynamic configuration:** An extension might need to fetch a configuration file to decide which listener options to set up (e.g. `webRequest` URL filters).

> 

Reviewer: Here is another use case, based on https://bugzilla.mozilla.org/show_bug.cgi?id=1822735 (starting at the fourth paragraph, starting with "The more serious issue"):

```suggestion
* **Atomic startup completion signal**: An extension that partially failed to register its listeners at its first run can have another chance at the next startup, without being in an inconsistent state.

```

Feel free to include a link to the bug in the text. Having concrete examples with more text might sometimes make it easier to understand what a short note is about.

### Known Consumers

Extensions which require dynamic or conditionally registered event listeners based on asynchronous state, without resorting to unconditionally waking up the service worker for events they may not care about.

## Specification

### Schema

#### Manifest

```json
{
  "background": {
    "service_worker": "background.js",
    "async_initialization": true
  }
}
```

#### Runtime API

```typescript
namespace runtime {
  /**
   * Signals to the browser that asynchronous initialization is complete.
   * Transitions the service worker out of the initialization phase, replaces 
   * the previous run's persisted listeners with the newly registered ones, 

>    * and flushes the event queue.

Reviewer: Since the functionality is specifically for the background context, shall we specify that it is only available there? So that e.g. an extension popup or options page cannot call it. Or do you want to leave that possibility open?

```suggestion
   * and flushes the event queue.
   *
   * This method is only available to the background script context.
```

   */  
  export function markInitializationComplete(): void;
}
```

### Behavior

#### Initialization Lifecycle

* **Manifest opt-in:** Extensions declare `"async_initialization": true` nested inside the `"background"` key in their `manifest.json`.
* **Renderer-side queueing:** During the initialization phase, the browser routes incoming events that match the previously registered listeners to the service worker's renderer process, which queues them instead of immediately attempting to dispatch them to JavaScript callbacks (which might not be attached yet).
* **Event routing (union of old and new):** While in the initialization phase, the browser will route an event to the renderer queue if it matches *either* the old persisted routing state (from the previous run) or any new listeners currently being registered by the executing script.

> * **Listener registration:** Developers perform their asynchronous setup (e.g., reading from storage) and call `addListener()` to attach their desired JavaScript callbacks *before* calling the completion API.

Reviewer: Presumably the setup may still be synchronous, adding "potentially" to signal that.

```suggestion
* **Listener registration:** Developers perform their potentially asynchronous setup (e.g., reading from storage) and call `addListener()` to attach their desired JavaScript callbacks *before* calling the completion API.
```

> * **State commit & completion API:** The extension must call `browser.runtime.markInitializationComplete()` once its async setup is complete. At that moment:

Reviewer: Added "potentially" before "async" again.

```suggestion
* **State commit & completion API:** The extension must call `browser.runtime.markInitializationComplete()` once its potentially async setup is complete. At that moment:
```

	1. The old persisted routing state is **thrown away**.
	2. The new listeners currently registered are committed to the browser as the new persisted routing state for future wake-ups.

> * **Event flush:** Once `markInitializationComplete()` ends the initialization phase, the renderer flushes its queue, dispatching all held events to the newly registered JavaScript callbacks in FIFO order.

Reviewer: What is the expectation when this method is called while the script is still executing synchronously at startup?

My expectation is:
- that there are no queued events at that time (because scripts execute synchronously and is not interrupted/paused to receive events)
- that any events delivered immediately after the script finishes execution are dispatched immediately, without a queue.

This is bordering on the implementation detail, but to encourage extensions to call the method ASAP (e.g. by relying on synchronous data at startup (#703) to cache results where possible), I'd like to include a note along the lines of:

> When the initialization completion is signaled while the background script is still starting/executing, events are not queued and the usual event dispatch happens as soon as possible.

This note does not have hard guarantees but expresses the intended semantics.

What are your thoughts on that?

Author: Agreed that calling the method during the initial run should ideally not add extra queueing latency compared with the usual registration path.

I'll add a note along those lines -- but it seems like it belongs in "Implementation Notes"? I put it there, but let me know if you feel strongly about it.

## Implementation Notes

* The `async_initialization` manifest key allows the browser to know before spinning up the service worker that it should start queuing events *in the renderer*.
* The browser's event router must compute the union of the old persisted listener configurations and any newly evaluated configurations during startup to determine which events to forward to the initializing renderer.
* Implementations must ensure that enqueued tasks are flushed and dispatched normally upon calling `markInitializationComplete()`. If the initialization fails before the method is called, the state commit is aborted, and the previous run's listener state is preserved in the browser process.

## Future Work

> N/A

Reviewer: Potential future work: dispatch `runtime.onPerformanceWarning` event when events have been dropped. This would enable extensions to detect bugs. This event was already described in #456 and implemented in Firefox, documented at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onPerformanceWarning

Author: What do you mean by "events have been dropped"? Because of queue overflow / initialization timeout?

Completing initialization without re-registering a listener can lead to "events being dropped" in some sense, but that would be working as intended.

Reviewer: > What do you mean by "events have been dropped"? Because of queue overflow / initialization timeout?

Yes to both. If the extension is somehow in a deadlock (e.g. initialization depends on an event, but events are all suppressed), then knowing that (critical) events have been dropped would be useful.

If the browser is permitted to flush the queue and force the initialization when a timeout is reached, notifying extensions would enable them to detect bugs. Perhaps this should be one of the events that are not delayed once registered.

In theory a full queue could eventually trigger a crash and theoretically trigger crash reports that could integrate in the [Reporting API](https://developer.mozilla.org/en-US/docs/Web/API/Reporting_API). Only Chrome  recognizes the "crash-reporting" [endpoint](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Reporting-Endpoints) though.

> Completing initialization without re-registering a listener can lead to "events being dropped" in some sense, but that would be working as intended.

Indeed.

# Proposal: Asynchronous Listener Registration

**Summary**

An API proposal to allow extensions to explicitly delay the dispatch of extension API events to their background context during its initialization phase via a manifest opt-in field and a `markListenerRegistrationComplete()` API, enabling the asynchronous registration of event listeners. 

**Document Metadata**

**Author:** @AndreaOrru

**Sponsoring Browser:** Google Chrome

**Contributors:** @rdcronin

**Created:** 2026-04-27

**Related Issues:** N/A

> ## Motivation

Reviewer 2: I think it may be worth introducing a "non-goals" or "out of scope" section to provide more information about what this proposal intentionally does not cover. As @Rob--W mentioned [in this comment](https://github.com/w3c/webextensions/pull/989#discussion_r3170648798), "standard SW lifecycle events, web platform APIs events, [and] non-event" are out of scope. It's not directly called out, but non-background contexts are also out of scope. 

A concrete use case for the latter is an extension popups fetching data from storage before calling `downloads.open()`. The user activation is lost because of the async storage operation, which will cause the `downloads.open()` call fail even though the call would succeed if it was made synchronously.

* **Atomic startup completion signal**: An extension that partially failed to register its listeners at its first run can have another chance at the next startup, without being in an inconsistent state (e.g. https://bugzilla.mozilla.org/show_bug.cgi?id=1822735).

### Known Consumers

Extensions which require dynamic or conditionally registered event listeners based on asynchronous state, without resorting to unconditionally waking up the background context for events they may not care about.

## Specification

### Schema

#### Manifest

```json
{
  "background": {
    "service_worker": "background.js",
    "async_listener_registration": true
  }
}
```

#### Runtime API

```typescript
namespace runtime {
  /**
   * Signals to the browser that asynchronous initialization is complete.
   * Transitions the background context out of the initialization phase, replaces
   * the previous run's persisted listeners with the newly registered ones, 
   * and flushes the event queue.
   *
   * This method is only available to the background script context.
   */  
  export function markListenerRegistrationComplete(): void;
}
```

### Behavior

> #### Initialization Lifecycle

Reviewer 2: We should call out in this section that handling of and limits on the event queues up to individual user agents.

While the limits themselves may be undefined, we may want to define some reporting mechanisms for these cases similar to what @Rob--W suggested [in this comment](https://github.com/w3c/webextensions/pull/989/changes#r3171127476).

* **Event routing (union of old and new):** While in the initialization phase, the browser will route an event to the renderer queue if it matches *either* the old persisted routing state (from the previous run) or any new listeners currently being registered by the executing script.
* **Listener registration:** Developers perform their potentially asynchronous setup (e.g., reading from storage) and call `addListener()` to attach their desired JavaScript callbacks *before* calling the completion API.
* **State commit & completion API:** The extension must call `browser.runtime.markListenerRegistrationComplete()` once its potentially async setup is complete. At that moment:
	1. The old persisted routing state is **thrown away**.
	2. The new listeners currently registered are committed to the browser as the new persisted routing state for future wake-ups.
* **Event flush:** Once `markListenerRegistrationComplete()` ends the initialization phase, the renderer flushes its queue, dispatching all held events to the newly registered JavaScript callbacks in FIFO order.

> #### Example Flow

Reviewer 2: Building on my comment about [late registration](https://github.com/w3c/webextensions/pull/989/changes#r3211706372), this section should be updated to include mention of what how the extension and browser handle the value of `shouldListen` changing in storage. This mentioned in the paragraph immediately after the code sample, but so quickly that I didn't absorb the impact of this passage. 

I'd suggest adding another extension code sample showing the extension updating the value of `shouldListen` and registering an event listener, then describing how this feeds back into the browser's extension initialization process.

4. Because the routing state is now cleared, future `tabs.onUpdated` events will no longer wake up the background context. This implicitly cleans up the routing state without the developer needing to explicitly call `removeListener()` (see Workaround B) or accept unnecessary wake-ups on future events (see Workaround A).

#### Edge Cases

* **Initialization failure fallback:** If the initialization phase fails (e.g., the worker crashes or hits a timeout before the completion API is called), the browser will abort the state commit. It will keep the previous persisted routing state to prevent leaving the extension in a broken state.
* **Late registration:** It is permissible to register new listeners after the `markListenerRegistrationComplete()` API call. If an extension does so, developers must understand the following lifecycle:
	* The browser will persist these "late" listeners across background context terminations to successfully trigger a future wake-up.
	* On the subsequent wake-up, all relevant listeners must be re-registered. Calling `markListenerRegistrationComplete()` entirely replaces the routing state from the previous run.

> 	* Therefore, any late-registered listeners from a previous run will be removed from the registered listeners during the next run's state commit unless the extension explicitly re-registers them before calling the completion API.

Reviewer 2: IMO "late registration" is a notable aspect of this problem space. While this isn't the main way we expect extensions to (intentionally) register event listeners, I think it's more than an edge case. I'd suggest moving this content into a "Runtime Lifecycle" (or similarly named) section directly under "Behavior" and describing the related systems in more detail. For example, the late listener registration list should also be mentioned in the Initialization Lifecycle section.

I'm also concerned about scenarios where session event listeners are misinterpreted as being late registrations for init listeners. While I don't have a concrete use case in mind, I'm imagining scenarios where an extension registers temporary event listeners on events like `webNavigation.onCompleted` or `tabs.onUpdated` to get updates on events related to a specific task being performed. If these registrations occur in the course of an extension's typical execution, could the behavior described here result in a significant increase in the number of unnecessary background wakes? It seems like it may, even in situations where developers attempt to limit the scope of the listener using filters. 

Admittedly some extensions may use `unregister()` to clean up these temporary event listeners, but as I see it the history of async extension registration is working against us here. Since post-init event listeners haven't impacted an extension outside of the current background context's lifetime, I expect that many extensions won't have listener clean up logic because the platform "just took care of it" for them.

Author: I agree that "late registration" shouldn't be described only as an edge case. I would move this into its own expanded section as you recommend, but let's discuss the semantics first. :)

The concern about temporary/session-scoped listeners is valid, although it's not clear to me whether you're advocating for changing the semantics or clarifying the documentation. At a minimum, we should make it very clear that a listener registered after `markListenerRegistrationComplete()` can affect future wake-up routing, and that listeners intended to be session-scoped must be removed explicitly.

That said, it's worth noting that under the current semantics, the impact of an accidentally persisted late listener is bounded by the next successful initialization commit: if the listener is not re-registered before the next `markListenerRegistrationComplete()` call, it is removed from the persisted routing state. This can still cause an unnecessary wake-up, so it's worth documenting prominently, but it doesn't persist indefinitely unless the extension keeps re-registering it.

The alternative would be to make late registrations ephemeral/session-scoped by default. The main downside is that intentional runtime configuration changes become harder. If an extension enables a feature after initialization and registers a new listener, that listener would not wake up the extension after the background context is terminated. The extension would then need some other way to commit updated routing state.

An explicit `persistent` event listener option, like the one I mentioned in https://github.com/w3c/webextensions/pull/989#issuecomment-4392317207, seems desirable because it would give developers direct control over whether a listener should survive background termination. Still, I'd prefer framing that as future work rather than expanding the scope of this proposal.

For this proposal, I think the better default is to keep late registrations persistent, document the lifecycle and risks clearly, and call out explicit session-scoped/non-persistent listeners as future work. But I am open to the alternative if others feel strongly about it.
